app/assets/javascripts/tictactoe.js

```diff
@@ -1 +1,217 @@
+  var square_3 = $('td[data-x="2"][data-y="0"]').text()
+  var square_4 = $('td[data-x="0"][data-y="1"]').text()
+  var square_5 = $('td[data-x="1"][data-y="1"]').text()
+  var square_6 = $('td[data-x="2"][data-y="1"]').text()
+  var square_7 = $('td[data-x="0"][data-y="2"]').text()
+  var square_8 = $('td[data-x="1"][data-y="2"]').text()
+  var square_9 = $('td[data-x="2"][data-y="2"]').text()
+
+  var squares = [square_1,square_2,square_3,square_4,square_5,square_6,square_7,square_8,square_9]
+  var id = $('td').data('id')
+
+  if (id == undefined) {
+    $.post('/games', { 'state[]': squares })
+  } else {
+    $.ajax({
+      data: { 'state': squares },
+      method: "PATCH",
+      url: "/games/" + id
+    });
+  }
+}
+
+function saveGame() {
+  $('#save').on('click', function () {
+    autoSaveGame()
+  })
+}
+
+
+function previousGame() {
+  $('#previous').on('click', function(){
+    $.get('/games', function (data){
+      var lists = ""
+      for (i = 0; i < data['data'].length; i++) {
+        lists += '<button class="js-game" data-id=' + data["data"][i]['id'] + '>' + (data["data"][i]['id']) + '</button> <br>'
+      }
+      $('#games').html(lists)
+
+    }).done(function () {
+      $(".js-game").on("click", function () {
+        var id = $(this).data("id");
+        $.get("/games/" + id, function(data) {
+            var square_1 = $('td[data-x="0"][data-y="0"]').text(data['data']["attributes"]['state'][0])
+            var square_2 = $('td[data-x="1"][data-y="0"]').text(data['data']["attributes"]['state'][1])
+            var square_3 = $('td[data-x="2"][data-y="0"]').text(data['data']["attributes"]['state'][2])
+            var square_4 = $('td[data-x="0"][data-y="1"]').text(data['data']["attributes"]['state'][3])
+            var square_5 = $('td[data-x="1"][data-y="1"]').text(data['data']["attributes"]['state'][4])
+            var square_6 = $('td[data-x="2"][data-y="1"]').text(data['data']["attributes"]['state'][5])
+            var square_7 = $('td[data-x="0"][data-y="2"]').text(data['data']["attributes"]['state'][6])
+            var square_8 = $('td[data-x="1"][data-y="2"]').text(data['data']["attributes"]['state'][7])
+            var square_9 = $('td[data-x="2"][data-y="2"]').text(data['data']["attributes"]['state'][8])
+
+            // this helps me in finding if the existing game exists or not. this will help me in sending a patch request or post request based on this id
+            $( "td" ).data('id', id)
+
+            // set the turn
+            let squares = []
+            squares = [square_1,square_2, square_3, square_4,square_5,square_6,square_7,square_8,square_9]
+
+            let texts = []
+            squares.forEach (function(e) {
+              return texts.push(e.text())
+            })
+
+            turn = 0
+            texts.forEach (function(e) {
+              if (e == "X" || e == "O") {
+	            turn++ }})
+
+      })
+    })
+  })
+})
+}
+
+function attachListeners () {
```

Comment:
It looks like you are calling `attachListeners` in your `document.ready`, but you have 1 or more listeners outside of that function.

```javascript
$(document).ready(function() {
  attachListeners();
})

function attachListeners() {
  $('td').on('click', function() {
    if (!$.text(this) && !checkWinner()) {
      doTurn(this);
    }
  });

  $('#save').on('click', () => saveGame());
  $('#previous').on('click', () => showPreviousGames());
  $('#clear').on('click', () => resetBoard());
}
```